scheduler: fix tracking of concurrent events

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The previous approach was based on the assumption that an in-flight pod can use the head of the received event list as marker for identifying all events that occur while the pod is in flight. That assumption is incorrect: when that existing element gets removed from the list because all pods that were in-flight when it was received are done, that marker's Next method returns nil and the code which should have seen several concurrent events (if there were any) missed all of those.

#### Which issue(s) this PR fixes:

A pod with concurrent events could incorrectly get moved to the unschedulable queue where it could got stuck until the next periodic purging after 5 minutes if there was no other event for it.


#### Special notes for your reviewer:

The approach with maintaining a single list of concurrent events can be fixed by inserting each in-flight pod into the list and using that element to identify "more recent" events for the pod.

This affects Kubernetes >= 1.28. We need to backport the fix.

#### Does this PR introduce a user-facing change?
```release-note
Fix a 1.28 regression in scheduler: a pod with concurrent events could incorrectly get moved to the unschedulable queue where it could got stuck until the next periodic purging after 5 minutes if there was no other event for it.
```

pohly committed:
scheduler: fix tracking of concurrent events

The previous approach was based on the assumption that an in-flight pod can use
the head of the received event list as marker for identifying all events that
occur while the pod is in flight. That assumption is incorrect: when that
existing element gets removed from the list because all pods that were
in-flight when it was received are done, that marker's Next method returns nil
and the code which should have seen several concurrent events (if there were
any) missed all of those.

As a result, a pod with concurrent events could incorrectly get moved to the
unschedulable queue where it could got stuck until the next periodic purging
after 5 minutes if there was no other event for it.

The approach with maintaining a single list of concurrent events can be fixed
by inserting each in-flight pod into the list and using that element to
identify "more recent" events for the pod.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -607,36 +614,48 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
+			logger.Error(nil, "In flight Pod isn't found in the scheduling queue. If you see this error log, it's likely a bug in the scheduler.", "pod", klog.KObj(pInfo.Pod))
+			return framework.QueueAfterBackoff
+		}
+		if p.inFlightEvents.Len() > len(p.inFlightPods) {
```
I am replicating the previous `if len(pInfo.UnschedulablePlugins) == 0` shortcut here, but only if absolutely necessary (= when pod is not in flight).

I am not exactly sure whether there are legitimate reasons for having it. One unit test depends on it, but perhaps that test is broken (see comment below).

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -1231,28 +1204,33 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
+	expectInFlightPods(t, q)
 	// Construct a Pod, but don't associate its scheduler failure to any plugin
-	hpp1 := highPriorityPodInfo.Pod.DeepCopy()
-	hpp1.Name = "hpp1"
```
Having two pods with the same UID breaks the `inFlightPod` tracking of pods. Somehow it worked before, but not anymore with the revised implementation.

alculquicondor replied:
good catch. We shouldn't be testing that.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -1231,28 +1204,33 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
 	if p, err := q.Pop(); err != nil || p.Pod != hpp1 {
 		t.Errorf("Expected: %v after Pop, but got: %v", hpp1, p.Pod.Name)
 	}
+	expectInFlightPods(t, q, hpp1.UID)
```
I used `expectInFlightPods` to figure out where the test flow went astray when it still used two pods with the same UID. I kept it because it provides some additional test coverage for both `inFlightPods` and `inFlightEvents` and because it makes the test easier to read.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -1263,32 +1241,41 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
 	}
+	expectInFlightPods(t, q, medPriorityPodInfo.Pod.UID, unschedulablePodInfo.Pod.UID, highPriorityPodInfo.Pod.UID, hpp1.UID)
 	q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(unschedulablePodInfo.Pod, "fooPlugin"), q.SchedulingCycle())
+	expectInFlightPods(t, q, medPriorityPodInfo.Pod.UID, highPriorityPodInfo.Pod.UID, hpp1.UID)
```
This is about the line above with `q.newQueuedPodInfo`: creating a new `PodInfo` instance feels wrong.

The test should more realistic and use the `podInfo` instance that it got from `Pop`, just like a real scheduling attempt.

I kept it because I don't want to rock the boat too much in one PR, but this might be worth a follow-up.

alculquicondor replied:
sg

pohly reviewed (commented):
/cc @Huang-Wei @alculquicondor

pohly commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -607,36 +614,48 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
+		// This can happen while updating a pod. In that case pInfo.UnschedulablePlugins should
+		// be empty. If it is not, we may have a problem.
+		if len(pInfo.UnschedulablePlugins) != 0 {
+			logger.Error(nil, "In flight Pod isn't found in the scheduling queue. If you see this error log, it's likely a bug in the scheduler.", "pod", klog.KObj(pInfo.Pod))
```
Code coverage and dlv show that this line is hit when running tests, for example:
```
  2669:				for _, op := range test.operations {
  2670:					for _, pInfo := range pInfos {
=>2671:						op(logger, queue, pInfo)
  2672:					}
  2673:				}
```

This is another instance of "`pInfo` is not the result of `Pop`". Doesn't look easy to fix, though.

pohly replied:
This code hits it reliably: https://github.com/kubernetes/kubernetes/blob/7e9fbc449ddccab5be18d7d5fa0a4158ec8227f2/pkg/scheduler/internal/queue/scheduling_queue_test.go#L2655

That's `addUnschedulablePodBackToBackoffQ` called for a fabricated `QueuedPodInfo`, without making sure that "inFlightPods" is set up. 

Current master also hits the error log message. This seems to be the only test which does, so it might be easier to fix than I thought.

alculquicondor replied:
ah I see. Yeah, we should make sure to add a Pop before calling unschedulable. I remember Kensei added a lot of those in several tests. These probably slipped.

pohly commented:
> But it only gets removed once that event no longer matters. Or is the calculation for "does it still matter" wrong?

The current definition of "does it still matter" is "the pods that where in flight when the event was received are still in flight". That's what the `PodInfo` counter is for.

What happens is this:
- pod1 popped
- pod2 popped
- event1 received -> stored as list head with count == 2
- pod3 popped, using that list head as `previousEvent`
- pod1 done -> counter == 1
- pod2 done -> counter == 0 -> element gets removed, which means that its `Next` always returns nil
- event2 received and relevant for in-flight pod3, gets stored as new list head with count == 1
- pod3 done ->  `previousEvent.Next` returns nil and thus event2 is not getting checked

alculquicondor commented:
> pod2 done -> counter == 0 -> element gets removed, which means that its Next always returns nil

I think the bug was on the fact that we assumed that Next would still work (point to the previous next). In any case, the new approach looks good.

However, let's try to get into why the pod UID is sometimes not found in `inFlightPods`

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -559,12 +517,27 @@ func Test_InFlightPods(t *testing.T) {
 
-			if diff := cmp.Diff(listToClusterEvents(test.wantReceivedEvents), listToClusterEvents(q.receivedEvents), cmp.AllowUnexported(clusterEvent{})); diff != "" {
-				t.Errorf("Unexpected diff in receivedEvents (-want, +got):\n%s", diff)
+			wantInFlightEvents := make([]interface{}, 0)
```
Then I need to make the same change in `listToValues` and remove the pre-allocation, otherwise I end up with `nil` != []interface{}` differences in some cases - done.

alculquicondor replied:
`cmpopts.EquateEmpty`

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -1306,6 +1293,35 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueue(t *testing.T) {
+	}
+	sortUIDs := cmpopts.SortSlices(func(a, b types.UID) bool { return a < b })
+	if diff := cmp.Diff(uids, actualUIDs, sortUIDs); diff != "" {
+		t.Fatalf("Unexpected content of inFlightPods (--- want, +++ have):\n%s", diff)
```
Changed.

pohly committed:
scheduler: unit test case for concurrent event with other pod

The problematic scenario was having one pod in flight, one event in the list,
and then detecting a concurrent event for a second pod after the first pod is
done. The new test case covers that.

To make it work without assumptions about the implementation, the QueuedPodInfo
returned by Pop must be the one passed to AddUnschedulableIfNotPresent
after (potentially) populating UnschedulablePlugins. This is done via callback
functions which bind to the same shared variable.

pohly commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -179,20 +179,12 @@ func TestPriorityQueue_AddWithReversePriorityLessFunc(t *testing.T) {
+	return values
 }
 
 func Test_InFlightPods(t *testing.T) {
```
I added one now. It's a bit more complicated, see second commit message:

> To make it work without assumptions about the implementation, the QueuedPodInfo
> returned by Pop must be the one passed to AddUnschedulableIfNotPresent
> after (potentially) populating UnschedulablePlugins. This is done via callback
> functions which bind to the same shared variable.

alculquicondor commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -523,6 +494,50 @@ func Test_InFlightPods(t *testing.T) {
+			wantActiveQPodNames: []string{pod2.Name},
+			wantInFlightPods:    nil,
+			wantInFlightEvents:  nil,
+			queueingHintMap: QueueingHintMapPerProfile{
```
I see that we didn't do this in other tests, but we should be listing outputs (`wantX`) last.

alculquicondor reviewed (commented):
/approve
/lgtm

pohly commented:
/hold

pull-kubernetes-linter-hints is right, my new code lacks errors checking:
```
ERROR: pkg/scheduler/internal/queue/scheduling_queue_test.go:508:36: Error return value of `q.AddUnschedulableIfNotPresent` is not checked (errcheck)
ERROR: 					q.AddUnschedulableIfNotPresent(logger, poppedPod, q.SchedulingCycle())
ERROR: 					                              ^
ERROR: pkg/scheduler/internal/queue/scheduling_queue_test.go:513:36: Error return value of `q.AddUnschedulableIfNotPresent` is not checked (errcheck)
ERROR: 					q.AddUnschedulableIfNotPresent(logger, poppedPod2, q.SchedulingCycle())
```

Fixed: https://github.com/kubernetes/kubernetes/compare/14ba163bd05d79db4315fea34cff9c367ff19159..c131c92b9f36f7fbd2b3625c3e1d0573932007b2

pohly commented:
/hold cancel

@alculquicondor: needs another LGTM.

alculquicondor commented:
/lgtm
Please prepare the cherry-pick

pohly commented:
Cherry-pick is in https://github.com/kubernetes/kubernetes/pull/120445
